### PR TITLE
docs: design doc for supporting sa authencation for RGW with vault

### PR DESCRIPTION
**Pull request by thotz**
**Description of your changes:**
The design doc for supporting service account authentication for RGW
while configuring with Vault. The OSD encryption already support it.

Signed-off-by: Jiffin Tony Thottan <jthottan@redhat.com>

**Which issue is resolved by this Pull Request:**
Resolves #

**Checklist:**

- [ ] **Commit Message Formatting**: Commit titles and messages follow guidelines in the [developer guide](https://rook.io/docs/rook/latest/development-flow.html#commit-structure).
- [ ] **Skip Tests for Docs**: If this is only a documentation change, add the label `skip-ci` on the PR.
- [ ] Reviewed the developer guide on [Submitting a Pull Request](https://rook.io/docs/rook/latest/development-flow.html#submitting-a-pull-request)
- [ ] [Pending release notes](https://github.com/rook/rook/blob/master/PendingReleaseNotes.md) updated with breaking and/or notable changes for the next minor release.
- [ ] Documentation has been updated, if necessary.
- [ ] Unit tests have been added, if necessary.
- [ ] Integration tests have been added, if necessary.

### CODE REVIEW

File: design/ceph/object/ceph-rgw-k8s-sa-authentication.md [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+### Non-Goals
+There are different ways RGW can authenticate with help of [vault agent](https://docs.ceph.com/en/latest/radosgw/vault/#vault-agent), but only service account authentication is supported here.
+
+## Proposal details
+The service account details will specified in `ConnectionDetails` of `KeyManagementServiceSpec`, using that info plus the other details a config map with <ceph-object-store>-vault-agent-cm populated for the vault agent sidecar container. The side container details will be added RGW pod spec and start sidecar container with RGW pod. The RGW will be configured with vault agent specific options. 
+
+### Risks and Mitigation
+User will be able to modify the `vault-agent-cm` which is not preferable
+
+## Alternatives
```
Comment (travisn) [inline design/ceph/object/ceph-rgw-k8s-sa-authentication.md:24]:
```suggestion
## Config Commands
```

**Review by travisn — CHANGES_REQUESTED** [review]

File: design/ceph/object/ceph-rgw-k8s-sa-authentication.md [hunk 2]
```diff
@@ -0,0 +1,29 @@
+---
+title: add service authentication with vault for RGW
+target-version: release-1.10
```
Comment (subhamkrai) [inline design/ceph/object/ceph-rgw-k8s-sa-authentication.md:3]:
```suggestion
target-version: release-1.10
```
?

File: design/ceph/object/ceph-rgw-k8s-sa-authentication.md [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+Configure RGW with service account authentication with Vault with help of vault agent.
+
+### Non-Goals
+There are different ways RGW can authenticate with help of [vault agent](https://docs.ceph.com/en/latest/radosgw/vault/#vault-agent), but only service account authentication is supported here.
+
+## Proposal details
+The service account details will specified in `ConnectionDetails` of `KeyManagementServiceSpec`, using that info plus the other details a config map with <ceph-object-store>-vault-agent-cm populated for the vault agent sidecar container. The side container details will be added RGW pod spec and start sidecar container with RGW pod. The RGW will be configured with vault agent specific options. 
+
+### Risks and Mitigation
+User will be able to modify the `vault-agent-cm` which is not preferable
```
[the remark this answers is not in the document]
Comment (thotz) [reply design/ceph/object/ceph-rgw-k8s-sa-authentication.md:22]: Rook Operator generates based on values from `Connection Details` in KMS config. please refer [here](https://github.com/rook/rook/pull/9872/files#diff-3217d32c0f1e6760077a7544b87ce13baceb1d6ce0e1931a72c69c6c1b14eaa9R768)

File: design/ceph/object/ceph-rgw-k8s-sa-authentication.md [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+When `Vault` is set up in Kubernetes environment, it can be authenticated using service account token. The OSD encryption already supports this feature and adding same for RGW. It uses sameset of  config option in `ConnectionDetails` for `KeyManagementServiceSpec`.
+
+### Goals
+Configure RGW with service account authentication with Vault with help of vault agent.
+
+### Non-Goals
+There are different ways RGW can authenticate with help of [vault agent](https://docs.ceph.com/en/latest/radosgw/vault/#vault-agent), but only service account authentication is supported here.
+
+## Proposal details
+The service account details will specified in `ConnectionDetails` of `KeyManagementServiceSpec`, using that info plus the other details a config map with <ceph-object-store>-vault-agent-cm populated for the vault agent sidecar container. The side container details will be added RGW pod spec and start sidecar container with RGW pod. The RGW will be configured with vault agent specific options.
```
[the remark this answers is not in the document]
Comment (thotz) [reply design/ceph/object/ceph-rgw-k8s-sa-authentication.md:19]: The configmap contains configurations for the start vault agent, I will add sample example here

File: design/ceph/object/ceph-rgw-k8s-sa-authentication.md [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+When `Vault` is set up in Kubernetes environment, it can be authenticated using service account token. The OSD encryption already supports this feature and adding same for RGW. It uses sameset of  config option in `ConnectionDetails` for `KeyManagementServiceSpec`.
+
+### Goals
+Configure RGW with service account authentication with Vault with help of vault agent.
+
+### Non-Goals
+There are different ways RGW can authenticate with help of [vault agent](https://docs.ceph.com/en/latest/radosgw/vault/#vault-agent), but only service account authentication is supported here.
+
+## Proposal details
+The service account details will specified in `ConnectionDetails` of `KeyManagementServiceSpec`, using that info plus the other details a config map with <ceph-object-store>-vault-agent-cm populated for the vault agent sidecar container. The side container details will be added RGW pod spec and start sidecar container with RGW pod. The RGW will be configured with vault agent specific options. For the following details in `ConnectionDetails`:
```
Comment (BlaineEXE) [inline design/ceph/object/ceph-rgw-k8s-sa-authentication.md:19]: Is this sidecar required? It is non-ideal for Rook to hard-code support for Vault, especially in today's world where the vault injector is available. Why don't we instruct users to use the injector instead?

**Review by BlaineEXE — CHANGES_REQUESTED** [review]
Is there a way for users to manually configure this today? If so, the design should also include those steps so we have an idea of what the operator needs to do to implement this?

What is the relative priority of this work compared to #10316 and #10318?

**Review by BlaineEXE — CHANGES_REQUESTED** [review]

File: design/ceph/object/ceph-rgw-k8s-sa-authentication.md (hunk 5 above)
Comment (thotz) [reply design/ceph/object/ceph-rgw-k8s-sa-authentication.md:19]: > Is this sidecar required? It is non-ideal for Rook to hard-code support for Vault, especially in today's world where the vault injector is available. Why don't we instruct users to use the injector instead?

Vault injector Job is different, it does not authenticate with applications. It just inject  vault secret directly to application pod as file kinda of webhook

**Comment by thotz** [comment]
> Is there a way for users to manually configure this today? If so, the design should also include those steps so we have an idea of what the operator needs to do to implement this?
> 
> What is the relative priority of this work compared to #10316 and #10318?

Please check [comment](https://github.com/rook/rook/pull/10318#issuecomment-1138763868)

**Review by BlaineEXE — CHANGES_REQUESTED** [review]
Requesting changes while this is waiting in priority line behind #10318 and #10323 (in that order).

**Comment by BlaineEXE** [comment]
I decided to spend my morning looking into Vault and SA-based Auth. I think the KMS and Kubernetes integration landscape has grown compared to when Vault support was initially added.


Of note, if users set up the kubernetes auth method to allow Vault to give KMS secrets to apps via service account, I don't see anything that suggests a Vault sidecar container is required. I think it will be best if we don't have to bind Vault-awareness into Rook more than we have to, and I think we should try to avoid creating a Vault pod (sidecar) if possible. While starting a vault agent on nodes may be the best strategy for deploying RGW on bare metal, I suspect there are more K-native methods available to us for Rook. We shouldn't mirror Ceph's complexity in Rook unless it's critical.

Vault's [agent injector](https://www.vaultproject.io/docs/platform/k8s/injector) exists, and I think this is a way we can add the vault sidecar without having to code the pod details into Rook. 

Or, I think there may be another option, which seems from my research today that it won't require a sidecar or the agent injector. From what I can tell, if the "Kubernetes" auth method is set up for Vault, and the Kubernetes version is 1.21+, a key will be added to `/var/run/secrets/kubernetes.io/serviceaccount` automatically. What I hope this means is that we can direct the RGW to find the auth details in that directory to authenticate with Vault.

These are the chief resources I found that break down the "Kubernetes" auth method:
- https://www.vaultproject.io/docs/auth/kubernetes#kubernetes-1-21
- https://kubernetes.io/docs/reference/access-authn-authz/service-accounts-admin/#serviceaccount-admission-controller

**Comment by thotz** [comment]
> I decided to spend my morning looking into Vault and SA-based Auth. I think the KMS and Kubernetes integration landscape has grown compared to when Vault support was initially added.
> 
> Of note, if users set up the kubernetes auth method to allow Vault to give KMS secrets to apps via service account, I don't see anything that suggests a Vault sidecar container is required.

For service account authentication vault sidecar is not a must. But here the issues RGW cannot directly authenticate with vault using the service account. But it can authenticate it with help of a vault agent. Hence vault agent is added as sidecar to rgw pod. So RGW authenticates with vault via vault agent using service account


 I think it will be best if we don't have to bind Vault-awareness into Rook more than we have to, and I think we should try to avoid creating a Vault pod (sidecar) if possible. While starting a vault agent on nodes may be the best strategy for deploying RGW on bare metal, I suspect there are more K-native methods available to us for Rook. We shouldn't mirror Ceph's complexity in Rook unless it's critical.
> 
> Vault's [agent injector](https://www.vaultproject.io/docs/platform/k8s/injector) exists, and I think this is a way we can add the vault sidecar without having to code the pod details into Rook.
> 
> Or, I think there may be another option, which seems from my research today that it won't require a sidecar or the agent injector. 

From what I can tell, if the "Kubernetes" auth method is set up for Vault, and the Kubernetes version is 1.21+, a key will be added to `/var/run/secrets/kubernetes.io/serviceaccount` automatically. What I hope this means is that we can direct the RGW to find the auth details in that directory to authenticate with Vault.

Yes that's correct and even I tried to add that support in RGW codebase https://github.com/ceph/ceph/pull/37868, this makes RGW understands the service account tokens of k8s. From design point of I felt the right approach to vault agent. At the these service account toke is jwt token which was already supported by RGW with vault agent.
> 
> These are the chief resources I found that break down the "Kubernetes" auth method:
> 
>     * https://www.vaultproject.io/docs/auth/kubernetes#kubernetes-1-21
> 
>     * https://kubernetes.io/docs/reference/access-authn-authz/service-accounts-admin/#serviceaccount-admission-controller

File: design/ceph/object/ceph-rgw-k8s-sa-authentication.md [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+   listener "tcp" {
+        address = "127.0.0.1:8100"
+        tls_disable = "true"
+   }
+```
+### Risks and Mitigation
+User will be able to modify the `vault-agent-cm` which is not preferable
+
+## Config Commands
+The user can itself bring up `vault agent` as separate deployment configuring with service account authentication details. Then set following for RGW via toolbox pod:
```
Comment (travisn) [inline design/ceph/object/ceph-rgw-k8s-sa-authentication.md:62]: If the vault agent can be started as a separate deployment, is only one agent needed for the whole cluster? If so, should we anyway create a vault agent deployment instead of an rgw sidecar? Or why should it be a sidecar for rgw?

**Review by travisn — CHANGES_REQUESTED** [review]

File: design/ceph/object/ceph-rgw-k8s-sa-authentication.md (hunk 6 above)
Comment (BlaineEXE) [reply design/ceph/object/ceph-rgw-k8s-sa-authentication.md:62]: Good question!

Comment (BlaineEXE) [reply design/ceph/object/ceph-rgw-k8s-sa-authentication.md:62]: I have done a fair bit more research on this. I believe that Vault's model that the vault agent runs as a sidecar is merely a requirement because it assumes the agent injects secrets into a shared pod volume. If RGW is talking to vault agent directly, there is no need to run the agent as a sidecar, because there is no need to have the agent inject anything into pod shared directories.

This means we can have (and capture in the design doc) the decision to either use a sidecar or to use a standalone vault agent that is shared by all RGWs. 

Using sidecars will naturally increase our resource footprint, a negative. However, the RGW can contact the sidecar on `localhost` within the Pod.

Using a standalone agent will decrease resources, especially for large numbers of RGWs. However, RGWs will likely be run on many nodes, and a shared agent will add additional host-to-host latency to each S3 call IIUC.

-----

Regarding the agent injector, while it is intended to be used to inject specific secrets into Pods, I believe it has the flexibility to configure the agent as a listener as well by modifying the ConfigMap as documented here: https://www.vaultproject.io/docs/platform/k8s/injector/examples#configmap-example

We can set the [listener config](https://www.vaultproject.io/docs/agent#listener) in the ConfigMap to have the agent sidecar automatically injected and listening on a port expected by the RGW.

In my opinion, a better design may be to allow users the flexibility to choose whether they want to run a standalone vault agent or whether they want sidecars. It would be easy to add an option to the CephObjectStore `security` spec that defines the address of the vault agent: `spec.security.vaultAgentHostname`.

If they want **sidecars**, then they can...
1. set `spec.security.vaultAgentHostname` to `localhost:<port>`
2. configure the agent `listener` to `localhost:<port>`
3. add the below annotations on the CephObjectStore.
    ```yaml
      annotations:
        vault.hashicorp.com/agent-inject: 'true'
        vault.hashicorp.com/agent-configmap: 'my-configmap' on the CephObjectStore and how to set the agent 
    ```
    
If they want to use a **standalone** vault agent, then they merely need to set `spec.security.vaultAgentHostname` to the Service that provides the vault agent.

Comment (BlaineEXE) [reply design/ceph/object/ceph-rgw-k8s-sa-authentication.md:62]: From the vault docs, `listener` specifies the address on which the vault agent listens. I believe this means we can (using the confimap) tell the agent that is injected to listen to RGW requests in the pod at something like `localhost:6543`. We can then configure `vaultAgentHostname` to be `localhost:6543`.

Comment (thotz) [reply design/ceph/object/ceph-rgw-k8s-sa-authentication.md:62]: @BlaineEXE: Yaa I can understand `listerner` with vault-agent, sorry I was confused about how it can be related to `vault-injector` The PR https://github.com/rook/rook/pull/9872https://github.com/rook/rook/pull/9872 have a similar configuration or approach

**Comment by travisn** [comment]
@thotz What's the status of this PR?

**Comment by thotz** [comment]
Closing this PR since I am not planning to work on it and there is very little interest for the feature

**Comment by travisn** [comment]
> If someone is interested to pick it up, we have already the design doc, proposed PR and tracker issue etc. @travisn do u way to track the features like this even if its closed, like new label or something??

A github issue should be opened if we still need to track the feature request.

**Comment by thotz** [comment]
@BlaineEXE @travisn  please review latest verison of this PR

File: design/ceph/object/ceph-rgw-k8s-sa-authentication.md [hunk 7]
```diff
@@ -0,0 +1,143 @@
+---
+Service authentication with vault for RGW
+target-version: release-1.14
```
Comment (travisn) [inline design/ceph/object/ceph-rgw-k8s-sa-authentication.md:3]: You're planning on implementing it immediately for the release soon?

File: design/ceph/object/ceph-rgw-k8s-sa-authentication.md [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+            config = {
+                role = "rook-ceph"
+            }
+        }
+   }
+   cache {
+        use_auto_auth_token = true
+   }
+   listener "tcp" {
+        address = "127.0.0.1:8100"
```
[the remark this answers is not in the document]
Comment (travisn) [reply design/ceph/object/ceph-rgw-k8s-sa-authentication.md:102]: ?

File: design/ceph/object/ceph-rgw-k8s-sa-authentication.md (hunk 7 above)
Comment (thotz) [reply design/ceph/object/ceph-rgw-k8s-sa-authentication.md:3]: Yes, thats the plan. The code changes for this implementation are pretty minimal. I will push the PR along with the design PR

File: design/ceph/object/ceph-rgw-k8s-sa-authentication.md [hunk 9]
```diff
@@ -0,0 +1,195 @@
...
+## Config Commands
+
+The user can itself bring up `vault agent` as separate deployment configuring with service account authentication details. Then set following for RGW via toolbox pod:
+
+```bash
+ceph config set <ceph auth user for rgw> rgw_crypt_sse_s3_vault_addr <vault agent address>
+ceph config set <ceph auth user for rgw> rgw_crypt_sse_s3_vault_auth agent
+```
+
+## Another approach
```
Comment (travisn) [inline design/ceph/object/ceph-rgw-k8s-sa-authentication.md:193]: If we're not considering this approach, I'd suggest to remove it from the design doc. There are lots of alternate ways to implement, but we don't need to list them.

File: design/ceph/object/ceph-rgw-k8s-sa-authentication.md [hunk 10]
```diff
@@ -0,0 +1,195 @@
...
+
+### Non-Goals
+
+There are different ways RGW can authenticate with help of [vault agent](https://docs.ceph.com/en/latest/radosgw/vault/#vault-agent), but only service account authentication is supported here.
+
+## Proposal details
+
+The [vault agent](https://developer.hashicorp.com/vault/docs/agent-and-proxy/agent#vault-stanza) can be configure with help of following annotation in `CephObjectStore CRD` :
+
+- vault.hashicorp.com/agent-inject --> brings vault agent with rgw pod
```
Comment (travisn) [inline design/ceph/object/ceph-rgw-k8s-sa-authentication.md:26]: Rook can generate these annotations on the rgw pod if needed, but why require them as annotations on the CRD? Rook always expects settings on the CRD, rather than annotations for settings.

File: design/ceph/object/ceph-rgw-k8s-sa-authentication.md [hunk 11]
```diff
@@ -0,0 +1,195 @@
...
+      connectionDetails:
+          KMS_PROVIDER: vault
+          VAULT_ADDR: http://127.0.0.1:8100
+          VAULT_BACKEND_PATH: rook
+          VAULT_SECRET_ENGINE: kv
+          VAULT_AUTH_METHOD: kubernetes
+          VAULT_AUTH_KUBERNETES_ROLE: rook-ceph
+
+---
+kind: ConfigMap
```
Comment (travisn) [inline design/ceph/object/ceph-rgw-k8s-sa-authentication.md:51]: Does the user create this configmap, or does Rook generate it? We need to be clear about what the admin needs to create, and what Rook will generate.

**Review by travisn — CHANGES_REQUESTED** [review]

File: design/ceph/object/ceph-rgw-k8s-sa-authentication.md (hunk 10 above)
Comment (thotz) [reply design/ceph/object/ceph-rgw-k8s-sa-authentication.md:26]: So will use settings in the CRD and add the annotation on RGW pod based on settings. Sounds good to me

File: design/ceph/object/ceph-rgw-k8s-sa-authentication.md (hunk 11 above)
Comment (thotz) [reply design/ceph/object/ceph-rgw-k8s-sa-authentication.md:51]: IMO since there are a lot of options I prefer the user to configure cm than by Rook Operator. We can provide the sample configuration(minimal configuration) as above

File: design/ceph/object/ceph-rgw-k8s-sa-authentication.md [hunk 12]
```diff
@@ -0,0 +1,182 @@
+---
+Service authentication with vault for RGW
+target-version: release-1.15
```
Comment (travisn) [inline design/ceph/object/ceph-rgw-k8s-sa-authentication.md:3]: Does #13934 already implement the complete design? Or how much is left? v1.15 is just in the next couple weeks, this seems too soon since it hasn't been reviewed yet.

**Review by travisn — COMMENTED** [review]
@thotz 
- Can you summarize the status?  Since it is such an old PR with infrequent updates, it's really hard to keep track of.
- Please summarize what has changed in the latest push.
- Let's not add the `keepalive` flag. That just causes us to forget about old PRs. At least the stale bot will remind us if the PR is not touched for a while.

**Comment by thotz** [comment]
> @thotz
> 
>     * Can you summarize the status?  Since it is such an old PR with infrequent updates, it's really hard to keep track of.
> 


>     * Please summarize what has changed in the latest push.


> 
>     * Let's not add the `keepalive` flag. That just causes us to forget about old PRs. At least the stale bot will remind us if the PR is not touched for a while.



> @thotz
> 
>     * Can you summarize the status?  Since it is such an old PR with infrequent updates, it's really hard to keep track of.
> 

I sent out code changes in https://github.com/rook/rook/pull/13934 as well. We set annotations in RGW service to bring up vault along with RGW server as sidecar. Through vault agent we can use service account authentication for vault encryption in RGW

>     * Please summarize what has changed in the latest push.

Earlier version PR(same for the code changes) annotations set directly on CephObjectStore gateway spec. After the discussion in last version, now Rook will set annotation based on options in `security.kms.connectionDetails`

> 
>     * Let's not add the `keepalive` flag. That just causes us to forget about old PRs. At least the stale bot will remind us if the PR is not touched for a while.

File: design/ceph/object/ceph-rgw-k8s-sa-authentication.md [hunk 13]
```diff
@@ -0,0 +1,182 @@
...
+The [vault agent](https://developer.hashicorp.com/vault/docs/agent-and-proxy/agent#vault-stanza) can be configure with help of following settings in `connectionDetails` in `CephObjectStore CRD` and it will set following annotations in rgw pods.
+
+* `VAULT_AGENT_ADDR`     --> provides address of vault agent
+* `VAULT_AGENT_CM`       --> if address is localhost then agent need to configured along with rgw pod passed with help of cm, role for k8s service account need to add here
+* `VAULT_AGENT_TLS_CERT` --> tls cert for vault agent to communicate with vault server
+
+If vault agent address points to localhost following annotation will set on rgw pod:
+
+* vault.hashicorp.com/agent-inject    --> brings vault agent with rgw pod
+* vault.hashicorp.com/agent-configmap -->  config map containing settings for vault agent pod
```
Comment (BlaineEXE) [inline design/ceph/object/ceph-rgw-k8s-sa-authentication.md:33]: I think it's okay if we allow users to set these annotations on the CephObjectStore themselves to inject the agent if they want the sidecar implementation.

But I do want to avoid Rook maintaining sidecar resource definitions.

File: design/ceph/object/ceph-rgw-k8s-sa-authentication.md [hunk 14]
```diff
@@ -0,0 +1,182 @@
...
+kind: Secret
+metadata:
+  name: <name of k8s secret>
+  namespace: rook-ceph
+type: Opaque
+data:
+  client.key: <>
+  client.crt: <>
+  ca.crt:     <>
+
```
Comment (BlaineEXE) [inline design/ceph/object/ceph-rgw-k8s-sa-authentication.md:154]: In this scenario, how does RGW know where to get the client ca/cert/key? Does it always load from `/vault/tls`?

**Review by BlaineEXE — COMMENTED** [review]
I get the sense that users could integrate with vault agent today in Rook without code chagnes by setting these annotations on the CephObjectStore ...

* vault.hashicorp.com/agent-inject    --> brings vault agent with rgw pod
* vault.hashicorp.com/agent-configmap -->  config map containing settings for vault agent pod
* vault.hashicorp.com/tls-secret  --> the name of kubernetes secret containing TLS certs to communicate with vault server, secret will be mounted in `/vault/tls`
* vault.hashicorp.com/ca-cert     --> path for ca cert `/vault/tls/ca.crt`
* vault.hashicorp.com/client-cert --> path for client cert from vault server  `/vault/tls/client.crt`
* vault.hashicorp.com/client-key  --> path for key from the vault server `/vault/tls/client.key`

... and then setting these configs using toolbox or override configmap ...

```
ceph config set <ceph auth user for rgw> rgw_crypt_sse_s3_vault_addr <vault agent address>
ceph config set <ceph auth user for rgw> rgw_crypt_sse_s3_vault_auth agent
```

Is that true? If so, then maybe the better thing to do would just be document that for users, and then plan our design more around APIs for the 2 `rgw_crypt_sse_s3_vault_*` settings.

**Comment by thotz** [comment]
> I get the sense that users could integrate with vault agent today in Rook without code chagnes by setting these annotations on the CephObjectStore ...
> 
>     * vault.hashicorp.com/agent-inject    --> brings vault agent with rgw pod
> 
>     * vault.hashicorp.com/agent-configmap -->  config map containing settings for vault agent pod
> 
>     * vault.hashicorp.com/tls-secret  --> the name of kubernetes secret containing TLS certs to communicate with vault server, secret will be mounted in `/vault/tls`
> 
>     * vault.hashicorp.com/ca-cert     --> path for ca cert `/vault/tls/ca.crt`
> 
>     * vault.hashicorp.com/client-cert --> path for client cert from vault server  `/vault/tls/client.crt`
> 
>     * vault.hashicorp.com/client-key  --> path for key from the vault server `/vault/tls/client.key`
> 
> 
> ... and then setting these configs using toolbox or override configmap ...
> 
> ```
> ceph config set <ceph auth user for rgw> rgw_crypt_sse_s3_vault_addr <vault agent address>
> ceph config set <ceph auth user for rgw> rgw_crypt_sse_s3_vault_auth agent
> ```
> 
> Is that true? If so, then maybe the better thing to do would just be document that for users, and then plan our design more around APIs for the 2 `rgw_crypt_sse_s3_vault_*` settings.

I have updated accordingly, currently only one option present for vault agent i.e `VAULT_AGENT_ADDR`

**Comment by BlaineEXE** [comment]
I may have to revisit this review. It looks like this PR already added most of the same functionality presented here, so I am really quite confused: https://github.com/rook/rook/pull/10204/files
